Add identifier to ORKRecorderConfiguration and ORKRecorder

rsanchezsaez opened the pull request:
[Issue #2](https://github.com/ResearchKit/ResearchKit/issues/2) implemented. 

Modeled after `ORKStep`'s `identifier` (read-only property that is assigned in the initializers).

The identifier gets passed from `ORKRecorderConfiguration` to `ORKRecorder` and finally to the corresponding `ORKFileResult`.

`ORKRecorderConfiguration` serialization added: `ORKTest` tests modified to support the new property and passing.

YuanZhu-apple commented on ResearchKit/ActiveTasks/ORKHealthQuantityTypeRecorder.m:
```diff
@@ -54,13 +54,15 @@ @interface ORKHealthQuantityTypeRecorder()
+                healthQuantityType:(HKQuantityType *)quantityType
                                       unit:(HKUnit *)unit
                                       step:(ORKStep *)step
                            outputDirectory:(NSURL *)outputDirectory
```
Align them with the first two lines.

YuanZhu-apple commented on ResearchKit/ActiveTasks/ORKAccelerometerRecorder.h:
```diff
@@ -49,10 +49,12 @@ ORK_CLASS_AVAILABLE
+- (instancetype)initWithIdentifier:(NSString *)identifer
+                         frequency:(double)frequency
                              step:(ORKStep *)step
                   outputDirectory:(NSURL *)outputDirectory;
```
Not aligned.

YuanZhu-apple commented on ResearchKit/ActiveTasks/ORKDeviceMotionRecorder.m:
```diff
@@ -50,11 +50,11 @@ @interface ORKDeviceMotionRecorder()
 {
-    self = [super initWithStep:step
+    self = [super initWithIdentifier:identifier step:step
                outputDirectory:(NSURL *)outputDirectory];
```
Align them in three lines?

YuanZhu-apple commented on ResearchKitTests/ORKRecorderTests.m:
```diff
@@ -366,10 +366,12 @@ - (void)checkResult {
-    recorder = [[ORKMockLocationRecorder alloc] initWithStep:recorder.step outputDirectory:recorder.outputDirectory];
+    recorder = [[ORKMockLocationRecorder alloc] initWithIdentifier:@"location"
+                                                              step:recorder.step
+                                             outputDirectory:recorder.outputDirectory];
```
Alignment.

YuanZhu-apple commented on ResearchKitTests/ORKRecorderTests.m:
```diff
@@ -414,13 +416,13 @@ - (void)testLocationRecorder {
+    ORKAccelerometerRecorderConfiguration *conf = [[ORKAccelerometerRecorderConfiguration alloc] initWithIdentifier:@"accelerometer" frequency:60.0];
     Class recorderClass = [ORKAccelerometerRecorder class];
     ORKAccelerometerRecorder *recorder = (ORKAccelerometerRecorder *)[self createRecorder:conf];
     
```
To see if the identifier get passed to recorder, can you add a check here?

YuanZhu-apple commented on ResearchKitTests/ORKRecorderTests.m:
```diff
@@ -530,7 +532,7 @@ - (void)testPedometerRecorder {
+    ORKTouchRecorder *recorder = (ORKTouchRecorder *)[self createRecorder:[[ORKTouchRecorderConfiguration alloc] initWithIdentifier:@"touch"]];
     
     XCTAssertTrue([recorder isKindOfClass:recorderClass], @"");
     
```
Also, add a check in `checkResult` to verify the result's identifier is correct.

YuanZhu-apple commented:
Overall, this is very good!

rsanchezsaez commented:
Thanks for your feedback! I added some commits fixing the issues.

YuanZhu-apple commented on Testing/ORKTest/ORKTestTests/ORKJSONSerializationTests.m:
```diff
@@ -220,7 +218,10 @@ - (void)testTaskModel {
     activeStep.stepDuration = 100.0;
-    activeStep.recorderConfigurations = @[[[ORKAccelerometerRecorderConfiguration alloc] initWithFrequency:11.0], [ORKTouchRecorderConfiguration new] , [ORKAudioRecorderConfiguration new]];
+    activeStep.recorderConfigurations =
+  @[[[ORKAccelerometerRecorderConfiguration alloc] initWithIdentifier:@"id.accelerometer" frequency:11.0],
```
Indentation.

rsanchezsaez commented:
Oops, missed that one. New commit pushed.

rsanchezsaez commented:
As per the *ResearchKit Contribution Terms and License Policy*, should I add my copyright to the files with more than 10 lines changed? Maybe just `ORKRecorder.[hm]`.

YuanZhu-apple commented on ResearchKit/ActiveTasks/ORKAccelerometerRecorder.h:
```diff
@@ -49,12 +49,14 @@ ORK_CLASS_AVAILABLE
  
+ @param identifier   The unique indentifier of the recorder (assigned by the recorder configuration).
  @param frequency    The frequency of accelerometer data collected from CoreMotion, in hertz (Hz).
  @return An initialized accelerometer recorder.
```
Can you also document the other parameters?

@param step    The step that requested this recording.
@param outputDirectory    The directory in which the _______ should be stored.

rsanchezsaez commented:
I added the missing initializer arguments to the documentation. I also added my copyright line to `ORKRecorder.[hm]`, tell me if you prefer not to have it.

Cheers!

YuanZhu-apple commented on ResearchKitTests/ORKRecorderTests.m:
```diff
@@ -331,16 +333,18 @@ - (void)tearDown {
+- (ORKRecorder *)createRecorder:(ORKRecorderConfiguration *)recorderConfiguration {
+    ORKRecorder *recorder = [recorderConfiguration recorderForStep:[[ORKStep alloc] initWithIdentifier:@"step"]
                                                                     outputDirectory:[NSURL fileURLWithPath:_outputPath]];
     recorder.delegate = self;
```
Would be nice to add a test here to verify the identifier get passed from `ORKRecorderConfiguration` to `recorder`.

YuanZhu-apple commented:
Your copyright line looks good.

rsanchezsaez commented:
Good idea, added.

jwe-apple commented on ResearchKit/ActiveTasks/ORKAccelerometerRecorder.m:
```diff
@@ -192,18 +190,27 @@ @implementation ORKAccelerometerRecorderConfiguration
-- (instancetype)initWithFrequency:(double)freq {
-    self = [self ork_init];
+
+- (instancetype)initWithIdentifier:(NSString *)identifier {
```
If it's really invalid to call this, this initializer should be marked NS_UNAVAILABLE in the header.

rsanchezsaez replied:
If we move `ORKRecorderConfiguration` `initWithIdentifier:(NSString *)identifier` from the *public* to the *internal* header (as per you suggestion in the comment below), do we still need to mark these as `NS_UNAVAILABLE`? They will just not be visible from outside the framework.

jwe-apple replied:
No, I think this is fine.

jwe-apple commented on ResearchKit/ActiveTasks/ORKDeviceMotionRecorder.m:
```diff
@@ -186,23 +188,29 @@ @implementation ORKDeviceMotionRecorderConfiguration
+    return [[ORKDeviceMotionRecorder alloc] initWithIdentifier:self.identifier
+                                                     frequency:self.frequency
+                                                          step:step
+                                               outputDirectory:(NSURL *)outputDirectory];
```
not your code, but why does this need a cast?

rsanchezsaez replied:
It doesn't, it's a common mistake I have spotted around. It happens when you copy the method declaration and paste it elsewhere to call it but then forget to delete some argument types. ;)

I'll remove the cast.

jwe-apple commented on ResearchKit/ActiveTasks/ORKRecorder.h:
```diff
@@ -58,19 +59,47 @@ ORK_CLASS_AVAILABLE
+ @return An initialized recorder configuration.
+ */
+- (instancetype)initWithIdentifier:(NSString *)identifier NS_DESIGNATED_INITIALIZER;
+- (instancetype)initWithCoder:(NSCoder *)aDecoder NS_DESIGNATED_INITIALIZER;
```
These initializers should not be exposed in the public header.

jwe-apple commented on ResearchKit/Common/ORKOrderedTask.m:
```diff
@@ -296,14 +296,33 @@ - (id)initWithCoder:(NSCoder *)aDecoder {
+static NSString * const ORKFitnessWalkPedometerRecorderIdentifier = @"fitness.walk.pedometer";
+static NSString * const ORKFitnessWalkAccelerometerRecorderIdentifier = @"fitness.walk.accelerometer";
+static NSString * const ORKFitnessWalkDeviceMotionRecorderIdentifier = @"fitness.walk.deviceMotion";
+static NSString * const ORKFitnessWalkLocationRecorderIdentifier = @"fitness.walk.location";
```
In my opinion, and I'm open to counter-argument, the identifiers here do not need prefixes (e.g. "fitness.walk."). We don't use prefixes on step identifiers, so there should be no need to do so at the next level of the hierarchy.

The hierarchy itself gives this context, and uniqueness is manually preserved because we only include one recorder of each type in a given step.

I would be in favor of having just one identifier name used for each type of recorder for these predefined recorders.

rsanchezsaez replied:
You make a very good point, I'm modifying this.

One question though, hypothetically couldn't several `ORKHealthQuantityTypeRecorder` be added to the same step (for different health units)?

jwe-apple commented on ResearchKit/Common/ORKOrderedTask.m:
```diff
@@ -296,14 +296,33 @@ - (id)initWithCoder:(NSCoder *)aDecoder {
 static NSString * const ORKInstruction0StepIdentifier = @"instruction";
```
This file now has many changes. You should add your copyright.

jwe-apple commented:
btw I'm sorry if my comments conflict with Yuan's in any way ;-)

rsanchezsaez commented:
By the way, one general style question. At which column number do you usually do a code line break? I'm specially asking for long initializers or method calls.

rsanchezsaez commented:
I have added commits covering your feedback (also check my replies to some of the comments). Let me know if there's anything else.

DanKeen commented:
Re: line breaks - wherever you think they feel right. Generally somewhere between 80 and 140. Folks have large monitors but we find that force-wrapped code still feels right to us old-timers.

rsanchezsaez commented:
@DanKeen Thanks for the info. I usually hard-wrap at 120 for my personal projects, but I was noticing that most of *ReasearchKit* code has longer lines (in many cases quite longer than 140).

jwe-apple commented:
Tested on iPhone 6 with ORKCatalog. All builds are clean. Looks good.
